openssl: better check that withPerl=false prevents perl reference

thomasjm opened the pull request:
###### Motivation for this change

This is a PR to fix #19965, making it possible to remove a large 50MB `perl` dependency from the closure of `openssl.bin` when using `withPerl=false`. The checks in `openssl/default.nix` weren't quite right before, and there was an unnecessarily restrictive assert preventing you from setting `withPerl` to `false` when not cross-compiling.

I wanted to do this so I could deploy `nodejs-slim` in a Docker container, without bringing in `perl`. I'll open a separate PR for that.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

thomasjm committed:
Make sure openssl.bin doesn't depend on perl when withPerl = false

veprbl commented on pkgs/development/libraries/openssl/default.nix:
```diff
@@ -163,6 +158,17 @@ let
         exit 1
       fi
+    '' + lib.optionalString (!withPerl) ''
+      # If withPerl is disabled, clear out any references to perl that remain in the bin output
+      # (after the postInstall hook above and fixup phase have run)
+      substituteInPlace $bin/bin/c_rehash --replace ${buildPackages.perl}/bin/perl "/usr/bin/env perl"
+      substituteInPlace $bin/bin/c_rehash --replace ${perl}/bin/perl "/usr/bin/env perl"
```
https://github.com/NixOS/nixpkgs/blob/83e00a588d25210f91af0cc55aa96220b0fc73d1/pkgs/development/libraries/openssl/default.nix#L57-L59
Can it be addressed here?

thomasjm replied:
You mean by customizing the interpreter path fixup hook somehow? I don't know, I've never done that before. I did this stuff in the fixup phase to avoid disrupting the earlier code path where a `postInstall` hook does a substitution for the interpreter path fixup hook to pick up.

veprbl replied:
By removing the last three lines. The `patchShebangs --build` should ignore perl after that.

veprbl commented on pkgs/development/libraries/openssl/default.nix:
```diff
@@ -163,6 +158,17 @@ let
+      substituteInPlace $bin/bin/c_rehash --replace ${buildPackages.perl}/bin/perl "/usr/bin/env perl"
+      substituteInPlace $bin/bin/c_rehash --replace ${perl}/bin/perl "/usr/bin/env perl"
+
+      # Check to make sure the bin output doesn't depend on perl
+      if grep -r '${buildPackages.perl}' $bin; then
+        echo "Found an erroneous dependency on perl in bin output ^^^" >&2
+        exit 1
+      fi
```
This is a copy-paste from above. Better be merged, or even better if `disallowedReferences` can be used.

thomasjm replied:
It's not quite a copy-paste because this one concerns the `bin` output rather than the main output.

I looked into `disallowedReferences` and it seems like it only works on main outputs. So I could use it on the first check, but then I'd still need a grep for the second check, which I think would be more confusing than the current code.

Not sure how to improve this while keeping the logic unchanged, any suggestions?

veprbl replied:
> It's not quite a copy-paste because this one concerns the `bin` output rather than the main output.

Even more reason to look into that, so that the difference is either highlighted or that the windows version might be outdated and should be using $bin as well.

> I looked into `disallowedReferences` and it seems like it only works on main outputs. So I could use it on the first check, but then I'd still need a grep for the second check, which I think would be more confusing than the current code.

This test shows that's not true:
```nix
with import <nixpkgs> {};

stdenv.mkDerivation {
  name = "test";

  outputs = [ "out" "foo" "bar" ];
  buildCommand = ''
    mkdir -p "$out" "$foo" "$bar"
    echo "${perl}" > "$bar/baz"
  '';

  disallowedReferences = [ perl ];
}
```
Fails with
```
error: output '/nix/store/gfsy34z2ikkj5hchdngi969qwmy574kr-test-bar' is not allowed to refer to the following paths:
         /nix/store/1xjwb0bd02n1y621mjqani68kh6n6gig-perl-5.34.0
```

thomasjm replied:
> This test shows that's not true

Sorry, I should have said it doesn't allow you to discriminate between different outputs, which presents similar problems for the current logic.

> Even more reason to look into that, so that the difference is either highlighted or that the windows version might be outdated

Well, I wanted to make this as low-impact as possible and not go changing the Windows behavior unnecessarily. But I just made a simplified version, let me push it and see what you think...

veprbl replied:
> Sorry, I should have said it doesn't allow you to discriminate between different outputs, which presents similar problems for the current logic.

Well, I assume your goal is to reduce the closure size, so you probably don't want the reference in any of the outputs.

> Well, I wanted to make this as low-impact as possible and not go changing the Windows behavior unnecessarily.

That's appreciated. I try to only suggest the simple things.

thomasjm commented on pkgs/development/libraries/openssl/default.nix:
```diff
@@ -6,14 +6,9 @@
 # This will cause c_rehash to refer to perl via the environment, but otherwise
 # will produce a perfectly functional openssl binary and library.
-, withPerl ? stdenv.hostPlatform == stdenv.buildPlatform
+, withPerl ? true
```
Note: I think it's confusing to flip `withPerl` based on whether you're cross-compiling or not. It seems clearer to just default to `true`.

veprbl replied:
Yeah, but then you need to take care to not break things. #108170 #82924

thomasjm replied:
Hmm okay, thanks for looking those up. I just put it back.

veprbl commented on pkgs/development/libraries/openssl/default.nix:
```diff
@@ -6,14 +6,9 @@
-assert (
-  lib.assertMsg (!withPerl -> stdenv.hostPlatform != stdenv.buildPlatform)
-  "withPerl should not be disabled unless cross compiling"
-);
```
👍

veprbl commented:
So at this point I see
```
# head $(nix-build -E '((import ./. {}).openssl.override { withPerl = false; }).bin')/bin/c_rehash
#!/usr/bin/env perl
```
So there is no bug, other than the pesky assert (the `withPerl` option introduced in ef24a2815ed949c86110339e048cef2b33b84297 and immediately locked out for no apparent reason). We could just remove `disallowedReferences` and merge this to master.

thomasjm commented:
Why not keep the `disallowedReferences` to keep this from regressing again? Looking at #19965 it seems like it has in the past.

veprbl commented:
> Why not keep the `disallowedReferences` to keep this from regressing again? Looking at #19965 it seems like it has in the past.

If the goal is to prevent dependency on `buildPackages.perl` spliced from the `nativeBuildInputs`, then it's probably need to be spelled out in `disallowedReferences` explicitly (not just `perl`) and unconditionally with respect to `withPerl`. At that point we might as well set `strictDeps = true;`.

veprbl commented:
Also can we organize the two changes into separate commits (or even better, separate PRs)?

thomasjm commented:
I only found a single passing reference to `strictDeps` in the nixpkgs manual which I don't fully understand, but happy to switch to it if you prefer. For now I just updated the `disallowedReferences`.

thomasjm commented:
I just rebased on `master`, squashed and force pushed since the other PR landed. Sorry if I stepped on what you were doing with `staging` @veprbl. Also sorry for the noise other reviewers, I don't seem to have the permissions to remove the pending reviewer requests.

jonringer commented on pkgs/development/libraries/openssl/default.nix:
```diff
@@ -152,13 +152,8 @@ let
-        exit 1
-      fi
-    '';
+    disallowedReferences = [buildPackages.perl]
```
Nit: most common way to express a list
```suggestion
    disallowedReferences = [ buildPackages.perl ]
```

Mindavi commented:
@ofborg eval

lf- commented:
This change breaks native compilation of openssl, since in the case of withPerl and native compilation, it will still have the buildPackages.perl path in the result as that is the same as `perl`:

```
dev/nixpkgs - [604c9732c75●] » nix build --print-out-paths .#perl
warning: Git tree '/home/jade/dev/nixpkgs' is dirty
/nix/store/76a9zh8vvgcm4szrwvgvg0bfpnj23p7p-perl-5.34.1-man
/nix/store/6qkn235ap89fpq0w1nr7x2ipf5a0f9cy-perl-5.34.1
dev/nixpkgs - [604c9732c75●] » nix build --print-out-paths .#buildPackages.perl
warning: Git tree '/home/jade/dev/nixpkgs' is dirty
/nix/store/76a9zh8vvgcm4szrwvgvg0bfpnj23p7p-perl-5.34.1-man
/nix/store/6qkn235ap89fpq0w1nr7x2ipf5a0f9cy-perl-5.34.1
```

The question this brings up is whether we should set withPerl to false always to avoid causing unnecessary closure size in native-built docker containers for instance. One potential unfortunate ramification of this is that on not nix-darwin/nixos systems, `/usr/bin/env perl` is probably someone else's perl. Is this catastrophic? Probably not; I don't think that script does anything terribly advanced anyway.

veprbl commented on pkgs/development/libraries/openssl/default.nix:
```diff
@@ -152,13 +152,8 @@ let
-        exit 1
-      fi
-    '';
+    disallowedReferences = [ buildPackages.perl ]
+      ++ lib.optional (!withPerl) perl;
```
Would this make sense?
```suggestion
    disallowedReferences = lib.optional ((buildPackages.perl != perl) || !withPerl) [ buildPackages.perl ];
```

lf- replied:
Not exactly; you always don't want build perl (assuming it's different from runtime perl), but you also don't want runtime perl if !withPerl

veprbl commented:
> Anything left to do on this PR? Would be great to have this merged.

A rebase would be a good start.

thomasjm commented:
Rebase done!

I think the solution in https://github.com/NixOS/nixpkgs/issues/19965#issuecomment-1242707773 is better though.

adrian-gierakowski commented:
> Rebase done!
> 
> I think the solution in [#19965 (comment)](https://github.com/NixOS/nixpkgs/issues/19965#issuecomment-1242707773) is better though.

Great, cause I already implemented it 😁
Will open a new PR soon. 

Should I drop the withPerl flag completely, or set it to false by default and use the replacement script and then allow people to set it to true if they do want the Perl based script for some reason?

thomasjm commented:
> Should I drop the withPerl flag completely

AFAICT from Googling around, the native version is preferred for modern OpenSSL, so I'd vote drop it.

adrian-gierakowski commented:
> > Should I drop the withPerl flag completely
> 
> AFAICT from Googling around, the native version is preferred for modern OpenSSL, so I'd vote drop it.

only noticed it now, I went for leaving it in with `false` as default: https://github.com/NixOS/nixpkgs/pull/219323
